django/db/models/expressions.py

```diff
@@ -536,6 +539,61 @@ def _prepare(self, output_field=None):
         return self
 
 
+class SlicableF(F):
+    """
+    An object that can add slicing notation to F expression.
+
+    Object resolves the column on which the slicing is applied, and then
+    applies the slicing if possible.
+    """
+    def __init__(self, f, k):
+        if not isinstance(k, (int, slice)):
+            raise TypeError
+        assert ((not isinstance(k, slice) and (k >= 0)) or
+                (isinstance(k, slice) and (k.start is None or k.start >= 0) and
+                 (k.stop is None or k.stop >= 0))), \
+            "Negative indexing is not supported."
+
+        assert ((not isinstance(k, slice)) or
+                (isinstance(k, slice) and k.step is None)), \
+            "Step argument is not supported."
+
+        self.expression = f
+        if isinstance(k, slice):
+            if k.start is not None:
+                self.low = int(k.start) + 1
+            else:
+                self.low = 1
+            if k.stop is not None:
+                self.length = int(k.stop) - self.low + 1
+            else:
+                self.length = None
+        else:
+            self.low = int(k)
+            self.length = 1
+
+    def __repr__(self):
+        return "{}({})".format(self.__class__.__name__, self.expression.name)
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        text_fields = ['BinaryField', 'CharField', 'EmailField', 'FilePathField',
+                       'GenericIPAddressField', 'SlugField', 'TextField', 'URLField'
+                       ]
+        resolved = query.resolve_ref(self.expression.name, allow_joins, reuse, summarize)
+        field_type = resolved.target.get_internal_type()
+        if field_type in text_fields:
+            from django.db.models.functions import Substr
+            return Substr(resolved, self.low, self.length)
+        elif field_type == "ArrayField":
+            from django.contrib.postgres.fields.array import SliceTransform
```

Comment:
core django shouldn't be importing a contrib module, this makes it impossible for third party extensions impossible to hook into this functionality (some of the fields in my library Django-MySQL would want to use this functionality). Maybe there should be a function on the field class that can implement the kind of transform to apply?

Reply:
Sorry I don't quite follow, a mixin for what class?

Reply:
Any class that derives from a new class called "Sliceable" can use the slicing operator. The fields can then be derived from this class as well to incorporate the required functionality.

It is also possible to add an attribute "is_sliceable"=False to the Field class. Any field that wants to allow for slicing can set this attribute to True. How would the slicing be implemented in that case is a question I don't have an answer to as yet.